src/components/elements/table/GenericTable.tsx

```diff
@@ -62,8 +73,14 @@ export interface Props<T> {
   >;
   filterToolbar?: ReactNode;
   noData?: ReactNode;
-  renderRow?: (row: T) => ReactNode;
+  renderRow?: (row: T, columnKeys: string[]) => ReactNode;
   condensed?: boolean;
+  // if overrideTableBody is true, GenericTable doesn't render a `tbody` element.
+  // This should only be used by tables that take over rendering using renderRow and render a `tbody` within their custom render fn
+  overrideTableBody?: boolean;
+  injectBelowRows?: ReactNode; // component to inject below all rendered rows, above footer
+  getTableRowActions?: (record: T) => TableRowActionsType;
```

Comment:
I'm wondering about the decision to make this a prop as opposed to just another column definition, can you say more about that? It does add more complexity to the already-complex GenericTable, so just wanting to think that through.

By a column definition, I mean:

```

const columns = [
         ...othercols,
        {
          key: 'actions',
          render: (enrollment) => (
            <TableRowActions
              record={enrollment}
              primaryAction={{
                title: 'View Enrollment',
                key: 'enrollment',
                ariaLabel: `View Enrollment, ${clientBriefName(enrollment.client)}`,
                to: generateSafePath(
                  EnrollmentDashboardRoutes.ENROLLMENT_OVERVIEW,
                  {
                    clientId: enrollment.client.id,
                    enrollmentId: enrollment.id,
                  }
                ),
              }}
              secondaryActions={[
                {
                  title: 'View Client',
                  key: 'client',
                  ariaLabel: `View Client, ${clientBriefName(enrollment.client)}`,
                  to: generateSafePath(ClientDashboardRoutes.PROFILE, {
                    clientId: enrollment.client.id,
                  }),
                },
              ]}
            />
          ),
        },
]


...
<GenericTableWithData columns={columns} />
```

Reply:
Looking at https://github.com/greenriver/hmis-frontend/pull/998/files, I wonder if this would also give us some more flexibility for use-cases when the action button is more complex. I haven't quite wrapped my head around everything yet, but would something like this simplify the changes in the other PR ?

Definitely an early thought, I'm sure there are things I'm missing about the implementation.



```js

const columns = [
         ...othercols,
        {
          key: 'actions',
          render: (enrollment) => (
            <TableRowActions
              record={enrollment}
              // use this prop if the primary action is a simple navigation
              primaryActionConfig={viewEnrollmentMenuItemConfig(enrollment}
              // use this prop if the primary action is more complicated, or if needs some special treatment
              renderPrimaryAction={<AssignServiceButton .... />}
              // maybe secondary actions are always expected to be navigation
              secondaryActionsConfig={[viewClientMenuItemConfig(enrollment)]}
            />
          ),
        },
]


...
<GenericTableWithData columns={columns} />
```

Reply:
I didn't think thoroughly about it, but it was based on the idea that, since this is a new pattern implemented consistently across many tables in the app, we want the common table component to guide usage and force consistency. 

Thinking beyond table row actions, the logical conclusion of that train of thought is that all table controls, like search, filter, toggle, add-record button, etc. live as props on the GenericTable. That doesn't seem desirable - it certainly would lead to a bloated GenericTable. But I guess forcing consistency is the one upside. Currently we rely on GenericTable callers to implement search bars and add-one buttons consistently every time, and there are places in the app where we aren't consistent. (For example, search and toggle on enrollments table vs. search on projects/orgs.)

Anyway, I guess those considerations seem out of scope... I don't feel strongly about it and can take a stab at implementing your suggestion.

Reply:
quick example of the consistency issue I'm mentioning, as actually applied to table row actions: https://github.com/greenriver/hmis-frontend/blob/martha/6825-hh-member-row/src/components/elements/table/GenericTable.tsx#L259-L261 By moving this out of GenericTable and into a regular column definition, we leave the column header (or lack thereof / accessibility thereof) up to the caller

Reply:
I've taken a shot at the refactor, I'd like to look at it together tomorrow: https://github.com/greenriver/hmis-frontend/compare/martha/6761-tables...martha/6761-tables-refactor?expand=1

Reply:
> we leave the column header (or lack thereof / accessibility thereof) up to the caller

This seems like something that should be addressed globally for all columns. The [ColumnDef](https://github.com/greenriver/hmis-frontend/blob/91e938c0daf2bfa026f2cc45de663d3e20c24812/src/components/elements/table/types.tsx#L8) has the `header` attribute as optional, so any column could be missing a header and become inaccessible. 

Looking at how we implement the Header cells:

https://github.com/greenriver/hmis-frontend/blob/91e938c0daf2bfa026f2cc45de663d3e20c24812/src/components/elements/table/GenericTable.tsx#L225-L236

it could be modified so that it sets a hidden header if no visible `header` is provided:
```
            <HeaderCell>
              {def.header ? (
                <strong>{def.header}</strong>
              ) : (
                <Box sx={visuallyHidden}>
                  {def.key || def.someOtherFieldThatsRequiredIfHeaderIsMissing}
                </Box>
              )}
            </HeaderCell>
```

That might address the concern more widely -- we could add additional attributes to ColumnDef if needed, to ensure that columns always have accessible headers.